cpu/stm32_common/periph/uart.c

```diff
@@ -145,7 +145,12 @@ static inline void wait_for_tx_complete(uart_t uart)
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     assert(uart < UART_NUMOF);
-
+    #if defined(DEVELHELP)
```

Comment:
I'm personally in favor of not having any white space in front of preprocessor directives, because:
 1. C and C preprocessor macros are two completely distinct languages and keeping the indent of the C code surrounding it cannot be applied consistently. See example below
 2. Most C code does this. I'm in favor for keeping conventions, unless there is a reason for not doing so
 3. The rest of this file does not use white space in front of preprocessor directives
 4. Preprocessor directives become more visible, if the have a zero indent compared to the C code. And you really don't want to overlook them :-)

Here an example with unclear level of indent.
```` C
int foo(enum bar, int blah)
{
    switch(bar) {
        #ifdef MODULE_FOO /* <-- two levels of indent */
        case some_enum_value:
            some_function(blah);
            break;
            #endif /* <-- two or three levels of indent? */
    }
}
````

However, in the RIOT code base there are both files that *do* *not* use whitespace before preprocessor directives, and files that *do* *use* whitespace. So there is apparently no rule on it. So treat this comment as my personal opinion and not as a review.

Reply:
You could also use `#ifdef DEVELHELP` here...

Reply:
Ya, I should have checked that.  I also made it just #if DEVELHELP, though I am not sure if that is preferred.

Reply:
> I'm personally in favor of not having any white space in front of preprocessor directives, because:

I'm in favour of avoiding preprocessor directives if possible:

```c
/* somewhere define a macro that is always defined */
#ifdef DEVELHELP
    #define DEVELHELP_ON 1
#else
    #define DEVELHELP_ON 0
#endif /* DEVELHELP */

/* then in the code you use a "real" if */

void uart_write(uart_t uart, const uint8_t *data, size_t len)
{
    assert(uart < UART_NUMOF);

    if (DEVELHELP_ON) {
        /* If tx is not enabled don't try to send */
         if (!(dev(uart)->CR1 & USART_CR1_TE)) {
             return;
         }
    }

/* rest of the function */
}
```

Isn't that much more readable? Also, you get the benefit of having the compiler SEE what is inside the if block always (even if it then gets thrown away.)

Reply:
> I'm in favour of avoiding preprocessor directives if possible

Let me point out that your code contains 5 preprocessor directives, which are 3 more than the original code. (It also adds 5 lines of code.)

I personally think the original code was more readable.

Reply:
I also agree with @maribu, and for the sake of uniformity.  I a bit think we are getting into a "lets delay this bugfix so we can be nit picky" territory.
...On that note I was thinking if it is a good idea to have something that would work if DEVELHELP==1 and crash if DEVELHELP==0.  Can we assume if DEVELHELP==0 the uart will always be initialized first?

Reply:
Just to make sure I did understand you correct: By "DEVELHELP==1" you mean `DEVELHELP` is defined, and "DEVELHELP==0" means `DEVELHELP` is not defined, right?

The only reason for a valid (`assert(uart < UART_NUMOF)` did not trigger) not to be initialized seems to be it is used early in the boot up process. The only use case to me seems debug output via `stdio`. In production there should be no debug output of the early boot process. (E.g. `sys/auto_init` will run only after the periph buses and `stdio` is available, so there is not to much code running before `stdio` is available.)

So I believe it is safe to assume that when `DEVELHELP` is not defined that UARTs are not used before they are initialized.